### PR TITLE
Migrate Yarn to PNPM no packages updates.

### PR DESCRIPTION
**Pull request by shrshindeMSFT**
For more information about how to contribute to this repo, visit [this page](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md).

## Description

As per Microsoft 1ES policy the support for `yarn` will be limited and will be deprecated in future. To address this issue, we are moving away from `yarn` to `pnpm` which is recommended by Microsoft 1ES. 

### What is pnpm and why pnpm?
`PNPM` is Performant NPM package manager which is faster, and space efficient package manager compared to `yarn 2`, `npm` or others. `PNPM` saves developer time to install dependencies again and again for each package/app in the mono-repo by having a virtual store on disk which is hard linked to local projects/mono-repos. As `pnpm` uses virtual store on the disk it saves developer tons of space and when installing dependencies, it is just linking to virtual store which makes it faster.

### What changes are required after this PR is merged?
1. Install `pnpm` on local machine via npm `npm install -g pnpm`. There are other ways to install `pnpm` which can be found [here.](https://pnpm.io/installation) 
2. Remove all the `node_modules`, run `yarn clean:full`. Or you can manually delete them from File Explorer/ Finder.
3. Installing dependencies `pnpm i` or `pnpm install`.
4. To build teams-js package run `pnpm build`.
5. For unit tests run `pnpm test`.
6. Start-test-app run `pnpm start-test-app`

Above are the commands most frequently used. Rest other commands can be found in root directory of the mono-repo in package.json.

Note: To generate docs you need to run `pnpm run docs`.

### Main changes in the PR:
Migration from `yarn` to `pnpm` package manager, some of the commands are modified and documented, in the `README` section. 

## Validation
Validated locally and on azure pipelines to build, test and run e2e tests as well.

### Unit Tests added:
Not required in this PR as the only change to the mono-repo is package manager. 

### End-to-end tests added:

N/A

### Change file added:

> Ensure the change file meets the [formatting requirements](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md#change-log-using-beachball).

Yes

### Related PRs:

> Remove this section if n/a

### Next/remaining steps:

Add a PR to app hosting sdk to adopt this change in the pipeline.

### CODE REVIEW

File: tools/cli/compareBundleAnalysis.js [hunk 1]
```diff
@@ -28,10 +28,10 @@ const { argv } = require('yargs')
   .option('baseBranchName', {
     string: true,
     demandOption: true,
-    description: 'Destination branch against which current Pull request\'s bundle size is compared',
+    description: "Destination branch against which current Pull request's bundle size is compared",
   });
 
-const { ADOSizeComparator, getAzureDevopsApi, bundlesContainNoChanges } = require('bundle-size-tools');
+const { ADOSizeComparator, getAzureDevopsApi, bundlesContainNoChanges } = require('../bundle-size-tools');
```
Comment (jadahiya-MSFT) [inline tools/cli/compareBundleAnalysis.js:34]: Just out of curiosity, why was this change needed?

**Review by jadahiya-MSFT — DISMISSED** [review]

File: azure-pipelines.yml [hunk 2]
```diff
@@ -58,11 +58,12 @@ jobs:
         condition: succeeded()
         workingDirectory: '$(AppHostingSdkProjectDirectory)'
 
-      - task: Yarn@2
+      - task: CMDLine@2
```
[the remark this answers is not in the document]
Comment (shrshindeMSFT) [reply azure-pipelines.yml:61]: In my testing the task has be consistent and no failures.

File: tools/cli/compareBundleAnalysis.js (hunk 1 above)
Comment (shrshindeMSFT) [reply tools/cli/compareBundleAnalysis.js:34]: As `yarn` build and store local packages in a `hoisted` form the `compareBundleAnalysis.js` script knows where the package is located. But `pnpm` uses hard links to resolve packages, we need to specify where the exact package is stored in the mono-repo.

File: apps/sample-app/webpack.config.js [hunk 3]
```diff
@@ -37,12 +37,14 @@ module.exports = {
...
-    publicPath: '/',
+    static: {
+      directory: path.join(__dirname, 'build'),
+      publicPath: '/',
+    },
     compress: true,
     port: 4003,
     https: true,
-    disableHostCheck: true,
+    allowedHosts: 'all',
```
Comment (Jessica-Koch) [inline apps/sample-app/webpack.config.js:47]: this may be super simple but but what does this do?

File: packages/teams-js/README.md [hunk 4]
```diff
@@ -20,7 +20,7 @@ To install the stable [version](https://learn.microsoft.com/javascript/api/overv
 
 ### yarn
```
Comment (Jessica-Koch) [inline packages/teams-js/README.md:21]: should this be pnpm?

**Review by Jessica-Koch — APPROVED** [review]

File: packages/teams-js/README.md (hunk 4 above)
Comment (shrshindeMSFT) [reply packages/teams-js/README.md]: Resolved in latest commit

**Review by Jessica-Koch — APPROVED** [review]

**Review by jadahiya-MSFT — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -44,7 +44,7 @@ Application used to test the functionality of the various library APIs.
 
 ### Locally generating reference documentation
 
-If you would like to locally generate reference documentation for TeamsJS v2, simply utilize the script `yarn docs` either from the monorepo root or inside the teams-js project root (`packages/teams-js`). This should output the generated documentation to `packages/teams-js/docs`.
+If you would like to locally generate reference documentation for TeamsJS v2, simply utilize the script `pnpm run docs` either from the monorepo root or inside the teams-js project root (`packages/teams-js`). This should output the generated documentation to `packages/teams-js/docs`.
```
Comment (erinha) [inline README.md:47]: Why do we have to use the `run` keyword for the `docs` script command?

File: azure-pipelines.yml [hunk 6]
```diff
@@ -81,11 +82,12 @@ jobs:
         parameters:
           appHostGitPath: git://$(AppHostingSdkGitPath3)@$(AppHostingSdkGitRef)
 
-      - task: Yarn@2
-        displayName: 'Build Test App Local'
+      - task: CmdLine@2
+        displayName: 'Build Test App local'
```
Comment (erinha) [inline azure-pipelines.yml:86]: nit: I feel "local" should remain capitalized

File: change/@microsoft-teams-js-d6954dd2-027c-4a76-b6b4-7f9eb86a815a.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "Move from Yarn to Pnpm",
```
Comment (erinha) [inline change/@microsoft-teams-js-d6954dd2-027c-4a76-b6b4-7f9eb86a815a.json:3]: Please use past tense, e.g., "Migrated from Yarn to pnpm for repository package management."

File: tools/cli/compareBundleAnalysis.js [hunk 8]
```diff
@@ -28,10 +28,10 @@ const { argv } = require('yargs')
   .option('baseBranchName', {
     string: true,
     demandOption: true,
-    description: 'Destination branch against which current Pull request\'s bundle size is compared',
+    description: "Destination branch against which current Pull request's bundle size is compared",
```
Comment (erinha) [inline tools/cli/compareBundleAnalysis.js:31]: nit: shouldn't this use single quotes?

File: README.md (hunk 5 above)
Comment (shrshindeMSFT) [reply README.md:47]: When `pnpm docs` is ran it just open a `https://www.npmjs.com/package/teams-js-monorepo` after debugging I found out that `typedoc` doesn't understand `pnpm` symlink to `typescript` therefore we need to use `pnpm run docs` instead, which helps `typedoc` to point to correct `typescript`.

File: azure-pipelines.yml (hunk 6 above)
Comment (shrshindeMSFT) [reply azure-pipelines.yml]: Resolved in latest commit.

**Review by erinha — APPROVED** [review]
🚀

**Review by Jessica-Koch — APPROVED** [review]
